[Issue]: Why rocPRIM use ::hip_bfloat16 as rocprim::bfloat16 instead of ::__hip_bfloat16？

### Problem Description

The thing is :

While I use `hipcub::DeviceSegmentedRadixSort::SortKeys<KeyT, int*>` with `KeyT == __hip_bfloat16`, then it leads to compile error which is 

```text
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/detail/device_radix_sort.hpp:196:46: error: no matching function for call to 'encode'
  196 |                 const bit_key_type bit_key = key_codec::encode(keys[i]);
      |                                              ^~~~~~~~~~~~~~~~~
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/detail/device_segmented_radix_sort.hpp:256:29: note: in instantiation of function template specialization 'rocprim::detail::radix_digit_count_helper<64, 256, 16, 8, false>::count_digits<false, const __hip_bfloat16 *, unsigned int>' requested here
  256 |         count_helper_type().count_digits(keys_input,
      |                             ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/detail/device_segmented_radix_sort.hpp:114:17: note: in instantiation of function template specialization 'rocprim::detail::segmented_radix_sort_helper<__hip_bfloat16, rocprim::empty_type, 64, 256, 16, 8, false>::sort<const __hip_bfloat16 *, __hip_bfloat16 *, rocprim::empty_type *, rocprim::empty_type *>' requested here
  114 |                 sort(keys_input,
      |                 ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/detail/device_segmented_radix_sort.hpp:934:38: note: in instantiation of function template specialization 'rocprim::detail::segmented_radix_sort_helper<__hip_bfloat16, rocprim::empty_type, 64, 256, 16, 8, false>::sort<const __hip_bfloat16 *, __hip_bfloat16 *, rocprim::empty_type *, rocprim::empty_type *>' requested here
  934 |             long_radix_helper_type().sort(keys_input,
      |                                      ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:129:9: note: in instantiation of function template specialization 'rocprim::detail::segmented_sort_large<rocprim::detail::target_config<rocprim::detail::wrapped_segmented_radix_sort_config<rocprim::default_config, __hip_bfloat16, rocprim::empty_type>, rocprim::detail::target_arch::gfx942>, false, const __hip_bfloat16 *, __hip_bfloat16 *, rocprim::empty_type *, rocprim::empty_type *, unsigned int *, int *>' requested here
  129 |         segmented_sort_large<decltype(arch_config), Descending>(keys_input,
      |         ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/config_types.hpp:454:28: note: in instantiation of function template specialization 'rocprim::detail::trampoline_kernel<rocprim::detail::wrapped_segmented_radix_sort_config<rocprim::default_config, __hip_bfloat16, rocprim::empty_type>, rocprim::detail::target_arch::gfx942, (lambda at /opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:127:19), rocprim::detail::default_config_selector>' requested here
  454 |             tuned_kernel = trampoline_kernel<Config, Arch, Kernel, LaunchSelector>;
      |                            ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/config_types.hpp:477:30: note: (skipping 1 context in backtrace; use -ftemplate-backtrace-limit=0 to see all)
  477 |     const auto launch_plan = make_launch_plan<Config, Kernel, LaunchSelector>(arch, kernel);
      |                              ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:144:12: note: in instantiation of function template specialization 'rocprim::detail::execute_launch_plan<rocprim::detail::wrapped_segmented_radix_sort_config<rocprim::default_config, __hip_bfloat16, rocprim::empty_type>, (lambda at /opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:127:19), rocprim::detail::default_config_selector>' requested here
  144 |     return execute_launch_plan<Config>(arch, kernel, grid, block, shmem, stream);
      |            ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:538:37: note: in instantiation of function template specialization 'rocprim::detail::launch_segmented_sort_large<rocprim::detail::wrapped_segmented_radix_sort_config<rocprim::default_config, __hip_bfloat16, rocprim::empty_type>, false, const __hip_bfloat16 *, __hip_bfloat16 *, rocprim::empty_type *, rocprim::empty_type *, unsigned int *, int *>' requested here
  538 |             ROCPRIM_RETURN_ON_ERROR(launch_segmented_sort_large<config, Descending>(
      |                                     ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:766:20: note: in instantiation of function template specialization 'rocprim::detail::segmented_radix_sort_impl<rocprim::default_config, false, const __hip_bfloat16 *, __hip_bfloat16 *, rocprim::empty_type *, rocprim::empty_type *, int *>' requested here
  766 |     return detail::segmented_radix_sort_impl<Config, false>(temporary_storage,
      |                    ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/hipcub/backend/rocprim/device/device_segmented_radix_sort.hpp:308:27: note: in instantiation of function template specialization 'rocprim::segmented_radix_sort_keys<rocprim::default_config, const __hip_bfloat16 *, __hip_bfloat16 *, int *, __hip_bfloat16>' requested here
  308 |         return ::rocprim::segmented_radix_sort_keys(d_temp_storage,
      |                           ^
main.cpp:64:48: note: in instantiation of function template specialization 'hipcub::DeviceSegmentedRadixSort::SortKeys<__hip_bfloat16, int *>' requested here
   64 |     status = hipcub::DeviceSegmentedRadixSort::SortKeys<KeyT, int*>(
      |                                                ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/../detail/../type_traits.hpp:880:29: note: candidate template ignored: couldn't infer template argument 'Decomposer'
  880 |         static bit_key_type encode(Key key, Decomposer decomposer = {})
      |                             ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/../detail/../type_traits.hpp:960:17: error: static assertion failed due to requirement 'sizeof(__hip_bfloat16) == 0': Only fundamental types (integral and floating point) are supported as radix sortkeys without a decomposer. For custom key types, use the extract_digit overloads with the decomposer argument
  960 |                 sizeof(bit_key_type) == 0,
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/detail/device_radix_sort.hpp:198:34: note: in instantiation of member function 'rocprim::traits::radix_key_codec::codec<__hip_bfloat16>::extract_digit' requested here
  198 |                     = key_codec::extract_digit(bit_key, bit, current_radix_bits);
      |                                  ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/detail/device_segmented_radix_sort.hpp:256:29: note: in instantiation of function template specialization 'rocprim::detail::radix_digit_count_helper<64, 256, 16, 8, false>::count_digits<false, const __hip_bfloat16 *, unsigned int>' requested here
  256 |         count_helper_type().count_digits(keys_input,
      |                             ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/detail/device_segmented_radix_sort.hpp:114:17: note: in instantiation of function template specialization 'rocprim::detail::segmented_radix_sort_helper<__hip_bfloat16, rocprim::empty_type, 64, 256, 16, 8, false>::sort<const __hip_bfloat16 *, __hip_bfloat16 *, rocprim::empty_type *, rocprim::empty_type *>' requested here
  114 |                 sort(keys_input,
      |                 ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/detail/device_segmented_radix_sort.hpp:934:38: note: in instantiation of function template specialization 'rocprim::detail::segmented_radix_sort_helper<__hip_bfloat16, rocprim::empty_type, 64, 256, 16, 8, false>::sort<const __hip_bfloat16 *, __hip_bfloat16 *, rocprim::empty_type *, rocprim::empty_type *>' requested here
  934 |             long_radix_helper_type().sort(keys_input,
      |                                      ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:129:9: note: in instantiation of function template specialization 'rocprim::detail::segmented_sort_large<rocprim::detail::target_config<rocprim::detail::wrapped_segmented_radix_sort_config<rocprim::default_config, __hip_bfloat16, rocprim::empty_type>, rocprim::detail::target_arch::gfx942>, false, const __hip_bfloat16 *, __hip_bfloat16 *, rocprim::empty_type *, rocprim::empty_type *, unsigned int *, int *>' requested here
  129 |         segmented_sort_large<decltype(arch_config), Descending>(keys_input,
      |         ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/config_types.hpp:454:28: note: (skipping 2 contexts in backtrace; use -ftemplate-backtrace-limit=0 to see all)
  454 |             tuned_kernel = trampoline_kernel<Config, Arch, Kernel, LaunchSelector>;
      |                            ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:144:12: note: in instantiation of function template specialization 'rocprim::detail::execute_launch_plan<rocprim::detail::wrapped_segmented_radix_sort_config<rocprim::default_config, __hip_bfloat16, rocprim::empty_type>, (lambda at /opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:127:19), rocprim::detail::default_config_selector>' requested here
  144 |     return execute_launch_plan<Config>(arch, kernel, grid, block, shmem, stream);
      |            ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:538:37: note: in instantiation of function template specialization 'rocprim::detail::launch_segmented_sort_large<rocprim::detail::wrapped_segmented_radix_sort_config<rocprim::default_config, __hip_bfloat16, rocprim::empty_type>, false, const __hip_bfloat16 *, __hip_bfloat16 *, rocprim::empty_type *, rocprim::empty_type *, unsigned int *, int *>' requested here
  538 |             ROCPRIM_RETURN_ON_ERROR(launch_segmented_sort_large<config, Descending>(
      |                                     ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:766:20: note: in instantiation of function template specialization 'rocprim::detail::segmented_radix_sort_impl<rocprim::default_config, false, const __hip_bfloat16 *, __hip_bfloat16 *, rocprim::empty_type *, rocprim::empty_type *, int *>' requested here
  766 |     return detail::segmented_radix_sort_impl<Config, false>(temporary_storage,
      |                    ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/hipcub/backend/rocprim/device/device_segmented_radix_sort.hpp:308:27: note: in instantiation of function template specialization 'rocprim::segmented_radix_sort_keys<rocprim::default_config, const __hip_bfloat16 *, __hip_bfloat16 *, int *, __hip_bfloat16>' requested here
  308 |         return ::rocprim::segmented_radix_sort_keys(d_temp_storage,
      |                           ^
main.cpp:64:48: note: in instantiation of function template specialization 'hipcub::DeviceSegmentedRadixSort::SortKeys<__hip_bfloat16, int *>' requested here
   64 |     status = hipcub::DeviceSegmentedRadixSort::SortKeys<KeyT, int*>(
      |                                                ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/../detail/../type_traits.hpp:960:38: note: expression evaluates to '2 == 0'
  960 |                 sizeof(bit_key_type) == 0,
      |                 ~~~~~~~~~~~~~~~~~~~~~^~~~
In file included from main.cpp:3:
In file included from /opt/rocm-7.1.0/lib/llvm/bin/../../../include/hipcub/hipcub.hpp:34:
In file included from /opt/rocm-7.1.0/lib/llvm/bin/../../../include/hipcub/backend/rocprim/hipcub.hpp:61:
In file included from /opt/rocm-7.1.0/lib/llvm/bin/../../../include/hipcub/backend/rocprim/device/device_radix_sort.hpp:38:
In file included from /opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_radix_sort.hpp:40:
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/detail/device_radix_sort.hpp:425:43: error: no matching function for call to 'get_out_of_bounds_key'
  425 |                 const Key out_of_bounds = key_codec::get_out_of_bounds_key();
      |                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/detail/device_segmented_radix_sort.hpp:270:35: note: in instantiation of function template specialization 'rocprim::detail::radix_sort_and_scatter_helper<256, 16, 8, false, __hip_bfloat16, rocprim::empty_type, unsigned int>::sort_and_scatter<false, const __hip_bfloat16 *, __hip_bfloat16 *, rocprim::empty_type *, rocprim::empty_type *>' requested here
  270 |         sort_and_scatter_helper().sort_and_scatter(keys_input,
      |                                   ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/detail/device_segmented_radix_sort.hpp:114:17: note: in instantiation of function template specialization 'rocprim::detail::segmented_radix_sort_helper<__hip_bfloat16, rocprim::empty_type, 64, 256, 16, 8, false>::sort<const __hip_bfloat16 *, __hip_bfloat16 *, rocprim::empty_type *, rocprim::empty_type *>' requested here
  114 |                 sort(keys_input,
      |                 ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/detail/device_segmented_radix_sort.hpp:934:38: note: in instantiation of function template specialization 'rocprim::detail::segmented_radix_sort_helper<__hip_bfloat16, rocprim::empty_type, 64, 256, 16, 8, false>::sort<const __hip_bfloat16 *, __hip_bfloat16 *, rocprim::empty_type *, rocprim::empty_type *>' requested here
  934 |             long_radix_helper_type().sort(keys_input,
      |                                      ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:129:9: note: in instantiation of function template specialization 'rocprim::detail::segmented_sort_large<rocprim::detail::target_config<rocprim::detail::wrapped_segmented_radix_sort_config<rocprim::default_config, __hip_bfloat16, rocprim::empty_type>, rocprim::detail::target_arch::gfx942>, false, const __hip_bfloat16 *, __hip_bfloat16 *, rocprim::empty_type *, rocprim::empty_type *, unsigned int *, int *>' requested here
  129 |         segmented_sort_large<decltype(arch_config), Descending>(keys_input,
      |         ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/config_types.hpp:454:28: note: in instantiation of function template specialization 'rocprim::detail::trampoline_kernel<rocprim::detail::wrapped_segmented_radix_sort_config<rocprim::default_config, __hip_bfloat16, rocprim::empty_type>, rocprim::detail::target_arch::gfx942, (lambda at /opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:127:19), rocprim::detail::default_config_selector>' requested here
  454 |             tuned_kernel = trampoline_kernel<Config, Arch, Kernel, LaunchSelector>;
      |                            ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/config_types.hpp:477:30: note: (skipping 1 context in backtrace; use -ftemplate-backtrace-limit=0 to see all)
  477 |     const auto launch_plan = make_launch_plan<Config, Kernel, LaunchSelector>(arch, kernel);
      |                              ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:144:12: note: in instantiation of function template specialization 'rocprim::detail::execute_launch_plan<rocprim::detail::wrapped_segmented_radix_sort_config<rocprim::default_config, __hip_bfloat16, rocprim::empty_type>, (lambda at /opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:127:19), rocprim::detail::default_config_selector>' requested here
  144 |     return execute_launch_plan<Config>(arch, kernel, grid, block, shmem, stream);
      |            ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:538:37: note: in instantiation of function template specialization 'rocprim::detail::launch_segmented_sort_large<rocprim::detail::wrapped_segmented_radix_sort_config<rocprim::default_config, __hip_bfloat16, rocprim::empty_type>, false, const __hip_bfloat16 *, __hip_bfloat16 *, rocprim::empty_type *, rocprim::empty_type *, unsigned int *, int *>' requested here
  538 |             ROCPRIM_RETURN_ON_ERROR(launch_segmented_sort_large<config, Descending>(
      |                                     ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:766:20: note: in instantiation of function template specialization 'rocprim::detail::segmented_radix_sort_impl<rocprim::default_config, false, const __hip_bfloat16 *, __hip_bfloat16 *, rocprim::empty_type *, rocprim::empty_type *, int *>' requested here
  766 |     return detail::segmented_radix_sort_impl<Config, false>(temporary_storage,
      |                    ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/hipcub/backend/rocprim/device/device_segmented_radix_sort.hpp:308:27: note: in instantiation of function template specialization 'rocprim::segmented_radix_sort_keys<rocprim::default_config, const __hip_bfloat16 *, __hip_bfloat16 *, int *, __hip_bfloat16>' requested here
  308 |         return ::rocprim::segmented_radix_sort_keys(d_temp_storage,
      |                           ^
main.cpp:64:48: note: in instantiation of function template specialization 'hipcub::DeviceSegmentedRadixSort::SortKeys<__hip_bfloat16, int *>' requested here
   64 |     status = hipcub::DeviceSegmentedRadixSort::SortKeys<KeyT, int*>(
      |                                                ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/../detail/../type_traits.hpp:1009:20: note: candidate function template not viable: requires single argument 'decomposer', but no arguments were provided
 1009 |         static Key get_out_of_bounds_key(Decomposer decomposer)
      |                    ^                     ~~~~~~~~~~~~~~~~~~~~~
In file included from main.cpp:3:
In file included from /opt/rocm-7.1.0/lib/llvm/bin/../../../include/hipcub/hipcub.hpp:34:
In file included from /opt/rocm-7.1.0/lib/llvm/bin/../../../include/hipcub/backend/rocprim/hipcub.hpp:61:
In file included from /opt/rocm-7.1.0/lib/llvm/bin/../../../include/hipcub/backend/rocprim/device/device_radix_sort.hpp:38:
In file included from /opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_radix_sort.hpp:40:
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/detail/device_radix_sort.hpp:447:17: error: no matching function for call to 'encode_inplace'
  447 |                 key_codec::encode_inplace(keys[i]);
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/../detail/../type_traits.hpp:896:21: note: candidate template ignored: couldn't infer template argument 'Decomposer'
  896 |         static void encode_inplace(Key& key, Decomposer decomposer = {})
      |                     ^
In file included from main.cpp:3:
In file included from /opt/rocm-7.1.0/lib/llvm/bin/../../../include/hipcub/hipcub.hpp:34:
In file included from /opt/rocm-7.1.0/lib/llvm/bin/../../../include/hipcub/backend/rocprim/hipcub.hpp:61:
In file included from /opt/rocm-7.1.0/lib/llvm/bin/../../../include/hipcub/backend/rocprim/device/device_radix_sort.hpp:38:
In file included from /opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_radix_sort.hpp:40:
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/detail/device_radix_sort.hpp:490:21: error: no matching function for call to 'decode_inplace'
  490 |                     key_codec::decode_inplace(key);
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/../detail/../type_traits.hpp:937:21: note: candidate template ignored: couldn't infer template argument 'Decomposer'
  937 |         static void decode_inplace(Key& key, Decomposer decomposer = {})
      |                     ^
In file included from main.cpp:3:
In file included from /opt/rocm-7.1.0/lib/llvm/bin/../../../include/hipcub/hipcub.hpp:34:
In file included from /opt/rocm-7.1.0/lib/llvm/bin/../../../include/hipcub/backend/rocprim/hipcub.hpp:61:
In file included from /opt/rocm-7.1.0/lib/llvm/bin/../../../include/hipcub/backend/rocprim/device/device_radix_sort.hpp:38:
In file included from /opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_radix_sort.hpp:40:
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/detail/device_radix_sort.hpp:196:46: error: no matching function for call to 'encode'
  196 |                 const bit_key_type bit_key = key_codec::encode(keys[i]);
      |                                              ^~~~~~~~~~~~~~~~~
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/detail/device_segmented_radix_sort.hpp:256:29: note: in instantiation of function template specialization 'rocprim::detail::radix_digit_count_helper<64, 256, 16, 8, false>::count_digits<false, __hip_bfloat16 *, unsigned int>' requested here
  256 |         count_helper_type().count_digits(keys_input,
      |                             ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/detail/device_segmented_radix_sort.hpp:141:17: note: in instantiation of function template specialization 'rocprim::detail::segmented_radix_sort_helper<__hip_bfloat16, rocprim::empty_type, 64, 256, 16, 8, false>::sort<__hip_bfloat16 *, __hip_bfloat16 *, rocprim::empty_type *, rocprim::empty_type *>' requested here
  141 |                 sort(keys_tmp,
      |                 ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/detail/device_segmented_radix_sort.hpp:934:38: note: in instantiation of function template specialization 'rocprim::detail::segmented_radix_sort_helper<__hip_bfloat16, rocprim::empty_type, 64, 256, 16, 8, false>::sort<const __hip_bfloat16 *, __hip_bfloat16 *, rocprim::empty_type *, rocprim::empty_type *>' requested here
  934 |             long_radix_helper_type().sort(keys_input,
      |                                      ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:129:9: note: in instantiation of function template specialization 'rocprim::detail::segmented_sort_large<rocprim::detail::target_config<rocprim::detail::wrapped_segmented_radix_sort_config<rocprim::default_config, __hip_bfloat16, rocprim::empty_type>, rocprim::detail::target_arch::gfx942>, false, const __hip_bfloat16 *, __hip_bfloat16 *, rocprim::empty_type *, rocprim::empty_type *, unsigned int *, int *>' requested here
  129 |         segmented_sort_large<decltype(arch_config), Descending>(keys_input,
      |         ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/config_types.hpp:454:28: note: in instantiation of function template specialization 'rocprim::detail::trampoline_kernel<rocprim::detail::wrapped_segmented_radix_sort_config<rocprim::default_config, __hip_bfloat16, rocprim::empty_type>, rocprim::detail::target_arch::gfx942, (lambda at /opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:127:19), rocprim::detail::default_config_selector>' requested here
  454 |             tuned_kernel = trampoline_kernel<Config, Arch, Kernel, LaunchSelector>;
      |                            ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/config_types.hpp:477:30: note: (skipping 1 context in backtrace; use -ftemplate-backtrace-limit=0 to see all)
  477 |     const auto launch_plan = make_launch_plan<Config, Kernel, LaunchSelector>(arch, kernel);
      |                              ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:144:12: note: in instantiation of function template specialization 'rocprim::detail::execute_launch_plan<rocprim::detail::wrapped_segmented_radix_sort_config<rocprim::default_config, __hip_bfloat16, rocprim::empty_type>, (lambda at /opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:127:19), rocprim::detail::default_config_selector>' requested here
  144 |     return execute_launch_plan<Config>(arch, kernel, grid, block, shmem, stream);
      |            ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:538:37: note: in instantiation of function template specialization 'rocprim::detail::launch_segmented_sort_large<rocprim::detail::wrapped_segmented_radix_sort_config<rocprim::default_config, __hip_bfloat16, rocprim::empty_type>, false, const __hip_bfloat16 *, __hip_bfloat16 *, rocprim::empty_type *, rocprim::empty_type *, unsigned int *, int *>' requested here
  538 |             ROCPRIM_RETURN_ON_ERROR(launch_segmented_sort_large<config, Descending>(
      |                                     ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:766:20: note: in instantiation of function template specialization 'rocprim::detail::segmented_radix_sort_impl<rocprim::default_config, false, const __hip_bfloat16 *, __hip_bfloat16 *, rocprim::empty_type *, rocprim::empty_type *, int *>' requested here
  766 |     return detail::segmented_radix_sort_impl<Config, false>(temporary_storage,
      |                    ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/hipcub/backend/rocprim/device/device_segmented_radix_sort.hpp:308:27: note: in instantiation of function template specialization 'rocprim::segmented_radix_sort_keys<rocprim::default_config, const __hip_bfloat16 *, __hip_bfloat16 *, int *, __hip_bfloat16>' requested here
  308 |         return ::rocprim::segmented_radix_sort_keys(d_temp_storage,
      |                           ^
main.cpp:64:48: note: in instantiation of function template specialization 'hipcub::DeviceSegmentedRadixSort::SortKeys<__hip_bfloat16, int *>' requested here
   64 |     status = hipcub::DeviceSegmentedRadixSort::SortKeys<KeyT, int*>(
      |                                                ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/../detail/../type_traits.hpp:880:29: note: candidate template ignored: couldn't infer template argument 'Decomposer'
  880 |         static bit_key_type encode(Key key, Decomposer decomposer = {})
      |                             ^
In file included from main.cpp:3:
In file included from /opt/rocm-7.1.0/lib/llvm/bin/../../../include/hipcub/hipcub.hpp:34:
In file included from /opt/rocm-7.1.0/lib/llvm/bin/../../../include/hipcub/backend/rocprim/hipcub.hpp:61:
In file included from /opt/rocm-7.1.0/lib/llvm/bin/../../../include/hipcub/backend/rocprim/device/device_radix_sort.hpp:38:
In file included from /opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_radix_sort.hpp:40:
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/detail/device_radix_sort.hpp:425:43: error: no matching function for call to 'get_out_of_bounds_key'
  425 |                 const Key out_of_bounds = key_codec::get_out_of_bounds_key();
      |                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/detail/device_segmented_radix_sort.hpp:270:35: note: in instantiation of function template specialization 'rocprim::detail::radix_sort_and_scatter_helper<256, 16, 8, false, __hip_bfloat16, rocprim::empty_type, unsigned int>::sort_and_scatter<false, __hip_bfloat16 *, __hip_bfloat16 *, rocprim::empty_type *, rocprim::empty_type *>' requested here
  270 |         sort_and_scatter_helper().sort_and_scatter(keys_input,
      |                                   ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/detail/device_segmented_radix_sort.hpp:141:17: note: in instantiation of function template specialization 'rocprim::detail::segmented_radix_sort_helper<__hip_bfloat16, rocprim::empty_type, 64, 256, 16, 8, false>::sort<__hip_bfloat16 *, __hip_bfloat16 *, rocprim::empty_type *, rocprim::empty_type *>' requested here
  141 |                 sort(keys_tmp,
      |                 ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/detail/device_segmented_radix_sort.hpp:934:38: note: in instantiation of function template specialization 'rocprim::detail::segmented_radix_sort_helper<__hip_bfloat16, rocprim::empty_type, 64, 256, 16, 8, false>::sort<const __hip_bfloat16 *, __hip_bfloat16 *, rocprim::empty_type *, rocprim::empty_type *>' requested here
  934 |             long_radix_helper_type().sort(keys_input,
      |                                      ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:129:9: note: in instantiation of function template specialization 'rocprim::detail::segmented_sort_large<rocprim::detail::target_config<rocprim::detail::wrapped_segmented_radix_sort_config<rocprim::default_config, __hip_bfloat16, rocprim::empty_type>, rocprim::detail::target_arch::gfx942>, false, const __hip_bfloat16 *, __hip_bfloat16 *, rocprim::empty_type *, rocprim::empty_type *, unsigned int *, int *>' requested here
  129 |         segmented_sort_large<decltype(arch_config), Descending>(keys_input,
      |         ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/config_types.hpp:454:28: note: in instantiation of function template specialization 'rocprim::detail::trampoline_kernel<rocprim::detail::wrapped_segmented_radix_sort_config<rocprim::default_config, __hip_bfloat16, rocprim::empty_type>, rocprim::detail::target_arch::gfx942, (lambda at /opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:127:19), rocprim::detail::default_config_selector>' requested here
  454 |             tuned_kernel = trampoline_kernel<Config, Arch, Kernel, LaunchSelector>;
      |                            ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/config_types.hpp:477:30: note: (skipping 1 context in backtrace; use -ftemplate-backtrace-limit=0 to see all)
  477 |     const auto launch_plan = make_launch_plan<Config, Kernel, LaunchSelector>(arch, kernel);
      |                              ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:144:12: note: in instantiation of function template specialization 'rocprim::detail::execute_launch_plan<rocprim::detail::wrapped_segmented_radix_sort_config<rocprim::default_config, __hip_bfloat16, rocprim::empty_type>, (lambda at /opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:127:19), rocprim::detail::default_config_selector>' requested here
  144 |     return execute_launch_plan<Config>(arch, kernel, grid, block, shmem, stream);
      |            ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:538:37: note: in instantiation of function template specialization 'rocprim::detail::launch_segmented_sort_large<rocprim::detail::wrapped_segmented_radix_sort_config<rocprim::default_config, __hip_bfloat16, rocprim::empty_type>, false, const __hip_bfloat16 *, __hip_bfloat16 *, rocprim::empty_type *, rocprim::empty_type *, unsigned int *, int *>' requested here
  538 |             ROCPRIM_RETURN_ON_ERROR(launch_segmented_sort_large<config, Descending>(
      |                                     ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_segmented_radix_sort.hpp:766:20: note: in instantiation of function template specialization 'rocprim::detail::segmented_radix_sort_impl<rocprim::default_config, false, const __hip_bfloat16 *, __hip_bfloat16 *, rocprim::empty_type *, rocprim::empty_type *, int *>' requested here
  766 |     return detail::segmented_radix_sort_impl<Config, false>(temporary_storage,
      |                    ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/hipcub/backend/rocprim/device/device_segmented_radix_sort.hpp:308:27: note: in instantiation of function template specialization 'rocprim::segmented_radix_sort_keys<rocprim::default_config, const __hip_bfloat16 *, __hip_bfloat16 *, int *, __hip_bfloat16>' requested here
  308 |         return ::rocprim::segmented_radix_sort_keys(d_temp_storage,
      |                           ^
main.cpp:64:48: note: in instantiation of function template specialization 'hipcub::DeviceSegmentedRadixSort::SortKeys<__hip_bfloat16, int *>' requested here
   64 |     status = hipcub::DeviceSegmentedRadixSort::SortKeys<KeyT, int*>(
      |                                                ^
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/../detail/../type_traits.hpp:1009:20: note: candidate function template not viable: requires single argument 'decomposer', but no arguments were provided
 1009 |         static Key get_out_of_bounds_key(Decomposer decomposer)
      |                    ^                     ~~~~~~~~~~~~~~~~~~~~~
In file included from main.cpp:3:
In file included from /opt/rocm-7.1.0/lib/llvm/bin/../../../include/hipcub/hipcub.hpp:34:
In file included from /opt/rocm-7.1.0/lib/llvm/bin/../../../include/hipcub/backend/rocprim/hipcub.hpp:61:
In file included from /opt/rocm-7.1.0/lib/llvm/bin/../../../include/hipcub/backend/rocprim/device/device_radix_sort.hpp:38:
In file included from /opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/device_radix_sort.hpp:40:
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/detail/device_radix_sort.hpp:447:17: error: no matching function for call to 'encode_inplace'
  447 |                 key_codec::encode_inplace(keys[i]);
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-7.1.0/lib/llvm/bin/../../../include/rocprim/device/../detail/../type_traits.hpp:896:21: note: candidate template ignored: couldn't infer template argument 'Decomposer'
  896 |         static void encode_inplace(Key& key, Decomposer decomposer = {})
      |                     ^
In file included from main.cpp:3:
In file included from /opt/rocm-7.1.0/lib/llvm/bin/../../../include/hipcub/hipcub.hpp:34:
In file included from /opt/rocm-7.1.0/lib/llvm/bin/../../../include/hipcub/backend/rocprim/hipcub.hpp:61:
fatal error: too many errors emitted, stopping now [-ferror-limit=]
5 warnings and 20 errors generated when compiling for gfx942.
```

then, I found that in file `rocprim/include/rocprim/types.hpp` use `hip_bfloat16`


<img width="545" height="110" alt="Image" src="https://github.com/user-attachments/assets/a3bf9490-4333-484f-b96e-4707b7a0b619" />

I think that ROCm recommend use `__hip_bflaot16`, but I use __hip_bfloat16 lead to compile error. So why is this situation? Will this be changed?




### Operating System

OS: NAME="Ubuntu" VERSION="24.04.3 LTS (Noble Numbat)"

### CPU

CPU:  model name      : INTEL(R) XEON(R) PLATINUM 8568Y+

### GPU

GPU:   Name:                    INTEL(R) XEON(R) PLATINUM 8568Y+      Marketing Name:          INTEL(R) XEON(R) PLATINUM 8568Y+      Name:                    gfx942                                Marketing Name:          AMD Instinct MI300X VF                    Name:                    amdgcn-amd-amdhsa--gfx942:sramecc+:xnack-       Name:                    amdgcn-amd-amdhsa--gfx9-4-generic:sramecc+:xnack-

### ROCm Version

ROCm 7.1.0

### ROCm Component

rocPRIM

### Steps to Reproduce

```C++
#include <hip/hip_runtime.h>
#include <hip/hip_bf16.h>
#include <hipcub/hipcub.hpp>
#include <hipcub/device/device_segmented_radix_sort.hpp>
#include <iostream>
#include <vector>

int main()
{
    using KeyT = __hip_bfloat16;
    // using KeyT = rocprim::bfloat16;     // 在rocprim中使用的是hip_bfloat16类型而不是__hip_bfloat16类型
    // hip_bfloat16
    // __hip_bfloat16

    // Host input keys in float for readability, then convert to bfloat16
    std::vector<float> h_floats{8.0f, 6.0f, 7.0f, 3.0f, 1.0f, 2.0f, 5.0f, 4.0f};
    const int num_items = static_cast<int>(h_floats.size());

    std::vector<KeyT> h_keys_in(num_items);
    for (int i = 0; i < num_items; ++i)
    {
        // h_keys_in[i] = __float2bfloat16(h_floats[i]);
        h_keys_in[i] = hip_bfloat16(h_floats[i]);
    }

    // Two segments: [0, 4) and [4, 8)
    std::vector<int> h_begin_offsets{0, 4};
    std::vector<int> h_end_offsets{4, 8};
    const int num_segments = static_cast<int>(h_begin_offsets.size());

    // Device memory
    KeyT* d_keys_in  = nullptr;
    KeyT* d_keys_out = nullptr;
    int* d_begin_offsets = nullptr;
    int* d_end_offsets   = nullptr;

    hipError_t status = hipSuccess;

    status = hipMalloc(&d_keys_in, num_items * sizeof(KeyT));
    if (status != hipSuccess) return status;
    status = hipMalloc(&d_keys_out, num_items * sizeof(KeyT));
    if (status != hipSuccess) return status;
    status = hipMalloc(&d_begin_offsets, num_segments * sizeof(int));
    if (status != hipSuccess) return status;
    status = hipMalloc(&d_end_offsets, num_segments * sizeof(int));
    if (status != hipSuccess) return status;

    status = hipMemcpy(d_keys_in, h_keys_in.data(),
                       num_items * sizeof(KeyT), hipMemcpyHostToDevice);
    if (status != hipSuccess) return status;
    status = hipMemcpy(d_begin_offsets, h_begin_offsets.data(),
                       num_segments * sizeof(int), hipMemcpyHostToDevice);
    if (status != hipSuccess) return status;
    status = hipMemcpy(d_end_offsets, h_end_offsets.data(),
                       num_segments * sizeof(int), hipMemcpyHostToDevice);
    if (status != hipSuccess) return status;

    // Temporary storage
    void*  d_temp_storage     = nullptr;
    size_t temp_storage_bytes = 0;

    // Step 1: query temporary storage size
    status = hipcub::DeviceSegmentedRadixSort::SortKeys<KeyT, int*>(
        d_temp_storage,
        temp_storage_bytes,
        d_keys_in,
        d_keys_out,
        num_items,
        num_segments,
        d_begin_offsets,
        d_end_offsets
    );
    if (status != hipSuccess) return status;

    // Allocate temporary storage
    status = hipMalloc(&d_temp_storage, temp_storage_bytes);
    if (status != hipSuccess) return status;

    // Step 2: perform segmented radix sort on bfloat16 keys
    status = hipcub::DeviceSegmentedRadixSort::SortKeys<KeyT, int*>(
        d_temp_storage,
        temp_storage_bytes,
        d_keys_in,
        d_keys_out,
        num_items,
        num_segments,
        d_begin_offsets,
        d_end_offsets
    );
    if (status != hipSuccess) return status;

    // Copy results back to host
    std::vector<KeyT> h_keys_out(num_items);
    status = hipMemcpy(h_keys_out.data(), d_keys_out,
                       num_items * sizeof(KeyT), hipMemcpyDeviceToHost);
    if (status != hipSuccess) return status;

    // Print results as float for readability
    std::cout << "Segmented radix sort (bfloat16 keys) result:\n";
    for (int i = 0; i < num_items; ++i)
    {
        // float value = __bfloat162float(h_keys_out[i]);
        auto value = float(h_keys_out[i]);
        std::cout << value << (i + 1 == num_items ? '\n' : ' ');
    }

    // Cleanup
    hipFree(d_temp_storage);
    hipFree(d_keys_in);
    hipFree(d_keys_out);
    hipFree(d_begin_offsets);
    hipFree(d_end_offsets);

    return 0;
}
```
```bash
hipcc main.cpp
```

### (Optional for Linux users) Output of /opt/rocm/bin/rocminfo --support

root@7:~/codes/rocprim-example# /opt/rocm/bin/rocminfo --support
ROCk module version 6.16.6 is loaded
=====================    
HSA System Attributes    
=====================    
Runtime Version:         1.18
Runtime Ext Version:     1.14
System Timestamp Freq.:  1000.000000MHz
Sig. Max Wait Duration:  18446744073709551615 (0xFFFFFFFFFFFFFFFF) (timestamp count)
Machine Model:           LARGE                              
System Endianness:       LITTLE                             
Mwaitx:                  DISABLED
XNACK enabled:           NO
DMAbuf Support:          YES
VMM Support:             YES

==========               
HSA Agents               
==========               
*******                  
Agent 1                  
*******                  
  Name:                    INTEL(R) XEON(R) PLATINUM 8568Y+   
  Uuid:                    CPU-XX                             
  Marketing Name:          INTEL(R) XEON(R) PLATINUM 8568Y+   
  Vendor Name:             CPU                                
  Feature:                 None specified                     
  Profile:                 FULL_PROFILE                       
  Float Round Mode:        NEAR                               
  Max Queue Number:        0(0x0)                             
  Queue Min Size:          0(0x0)                             
  Queue Max Size:          0(0x0)                             
  Queue Type:              MULTI                              
  Node:                    0                                  
  Device Type:             CPU                                
  Cache Info:              
    L1:                      32768(0x8000) KB                   
  Chip ID:                 0(0x0)                             
  ASIC Revision:           0(0x0)                             
  Cacheline Size:          64(0x40)                           
  Max Clock Freq. (MHz):   0                                  
  BDFID:                   0                                  
  Internal Node ID:        0                                  
  Compute Unit:            20                                 
  SIMDs per CU:            0                                  
  Shader Engines:          0                                  
  Shader Arrs. per Eng.:   0                                  
  WatchPts on Addr. Ranges:1                                  
  Memory Properties:       
  Features:                None
  Pool Info:               
    Pool 1                   
      Segment:                 GLOBAL; FLAGS: FINE GRAINED        
      Size:                    247409240(0xebf2a58) KB            
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Recommended Granule:4KB                                
      Alloc Alignment:         4KB                                
      Accessible by all:       TRUE                               
    Pool 2                   
      Segment:                 GLOBAL; FLAGS: EXTENDED FINE GRAINED
      Size:                    247409240(0xebf2a58) KB            
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Recommended Granule:4KB                                
      Alloc Alignment:         4KB                                
      Accessible by all:       TRUE                               
    Pool 3                   
      Segment:                 GLOBAL; FLAGS: KERNARG, FINE GRAINED
      Size:                    247409240(0xebf2a58) KB            
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Recommended Granule:4KB                                
      Alloc Alignment:         4KB                                
      Accessible by all:       TRUE                               
    Pool 4                   
      Segment:                 GLOBAL; FLAGS: COARSE GRAINED      
      Size:                    247409240(0xebf2a58) KB            
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Recommended Granule:4KB                                
      Alloc Alignment:         4KB                                
      Accessible by all:       TRUE                               
  ISA Info:                
*******                  
Agent 2                  
*******                  
  Name:                    gfx942                             
  Uuid:                    GPU-e23035068765b63a               
  Marketing Name:          AMD Instinct MI300X VF             
  Vendor Name:             AMD                                
  Feature:                 KERNEL_DISPATCH                    
  Profile:                 BASE_PROFILE                       
  Float Round Mode:        NEAR                               
  Max Queue Number:        128(0x80)                          
  Queue Min Size:          64(0x40)                           
  Queue Max Size:          131072(0x20000)                    
  Queue Type:              MULTI                              
  Node:                    1                                  
  Device Type:             GPU                                
  Cache Info:              
    L1:                      32(0x20) KB                        
    L2:                      4096(0x1000) KB                    
    L3:                      262144(0x40000) KB                 
  Chip ID:                 29877(0x74b5)                      
  ASIC Revision:           1(0x1)                             
  Cacheline Size:          128(0x80)                          
  Max Clock Freq. (MHz):   2100                               
  BDFID:                   33536                              
  Internal Node ID:        1                                  
  Compute Unit:            304                                
  SIMDs per CU:            4                                  
  Shader Engines:          32                                 
  Shader Arrs. per Eng.:   1                                  
  WatchPts on Addr. Ranges:4                                  
  Coherent Host Access:    FALSE                              
  Memory Properties:       
  Features:                KERNEL_DISPATCH 
  Fast F16 Operation:      TRUE                               
  Wavefront Size:          64(0x40)                           
  Workgroup Max Size:      1024(0x400)                        
  Workgroup Max Size per Dimension:
    x                        1024(0x400)                        
    y                        1024(0x400)                        
    z                        1024(0x400)                        
  Max Waves Per CU:        32(0x20)                           
  Max Work-item Per CU:    2048(0x800)                        
  Grid Max Size:           4294967295(0xffffffff)             
  Grid Max Size per Dimension:
    x                        2147483647(0x7fffffff)             
    y                        65535(0xffff)                      
    z                        65535(0xffff)                      
  Max fbarriers/Workgrp:   32                                 
  Packet Processor uCode:: 189                                
  SDMA engine uCode::      25                                 
  IOMMU Support::          None                               
  Pool Info:               
    Pool 1                   
      Segment:                 GLOBAL; FLAGS: COARSE GRAINED      
      Size:                    200998912(0xbfb0000) KB            
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Recommended Granule:2048KB                             
      Alloc Alignment:         4KB                                
      Accessible by all:       FALSE                              
    Pool 2                   
      Segment:                 GLOBAL; FLAGS: EXTENDED FINE GRAINED
      Size:                    200998912(0xbfb0000) KB            
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Recommended Granule:2048KB                             
      Alloc Alignment:         4KB                                
      Accessible by all:       FALSE                              
    Pool 3                   
      Segment:                 GLOBAL; FLAGS: FINE GRAINED        
      Size:                    200998912(0xbfb0000) KB            
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Recommended Granule:2048KB                             
      Alloc Alignment:         4KB                                
      Accessible by all:       FALSE                              
    Pool 4                   
      Segment:                 GROUP                              
      Size:                    64(0x40) KB                        
      Allocatable:             FALSE                              
      Alloc Granule:           0KB                                
      Alloc Recommended Granule:0KB                                
      Alloc Alignment:         0KB                                
      Accessible by all:       FALSE                              
  ISA Info:                
    ISA 1                    
      Name:                    amdgcn-amd-amdhsa--gfx942:sramecc+:xnack-
      Machine Models:          HSA_MACHINE_MODEL_LARGE            
      Profiles:                HSA_PROFILE_BASE                   
      Default Rounding Mode:   NEAR                               
      Default Rounding Mode:   NEAR                               
      Fast f16:                TRUE                               
      Workgroup Max Size:      1024(0x400)                        
      Workgroup Max Size per Dimension:
        x                        1024(0x400)                        
        y                        1024(0x400)                        
        z                        1024(0x400)                        
      Grid Max Size:           4294967295(0xffffffff)             
      Grid Max Size per Dimension:
        x                        2147483647(0x7fffffff)             
        y                        65535(0xffff)                      
        z                        65535(0xffff)                      
      FBarrier Max Size:       32                                 
    ISA 2                    
      Name:                    amdgcn-amd-amdhsa--gfx9-4-generic:sramecc+:xnack-
      Machine Models:          HSA_MACHINE_MODEL_LARGE            
      Profiles:                HSA_PROFILE_BASE                   
      Default Rounding Mode:   NEAR                               
      Default Rounding Mode:   NEAR                               
      Fast f16:                TRUE                               
      Workgroup Max Size:      1024(0x400)                        
      Workgroup Max Size per Dimension:
        x                        1024(0x400)                        
        y                        1024(0x400)                        
        z                        1024(0x400)                        
      Grid Max Size:           4294967295(0xffffffff)             
      Grid Max Size per Dimension:
        x                        2147483647(0x7fffffff)             
        y                        65535(0xffff)                      
        z                        65535(0xffff)                      
      FBarrier Max Size:       32                                 
*** Done ***             

### Additional Information

_No response_

--- a/projects/rocprim/CHANGELOG.md
+++ b/projects/rocprim/CHANGELOG.md
@@ -4,6 +4,10 @@ Full documentation for rocPRIM is available at [https://rocm.docs.amd.com/projec
 
 ## rocPRIM 4.4.0 for ROCm 7.13
 
+### Added
+
+* Added type trait definitions for `__hip_bfloat16`. This should resolve issues where this type did not work with radix-based algorithms.
+
 ### Changed
 
 * Building benchmarks on Windows is not currently possible because of the dependency on AMD SMI. 

--- a/projects/rocprim/rocprim/include/rocprim/type_traits.hpp
+++ b/projects/rocprim/rocprim/include/rocprim/type_traits.hpp
@@ -1,4 +1,4 @@
-// Copyright (c) 2024-2025 Advanced Micro Devices, Inc. All rights reserved.
+// Copyright (c) 2024-2026 Advanced Micro Devices, Inc. All rights reserved.
 //
 // Permission is hereby granted, free of charge, to any person obtaining a copy
 // of this software and associated documentation files (the "Software"), to deal
@@ -1154,15 +1154,23 @@ struct get
     /// float, double);
     constexpr bool is_build_in() const
     {
-        return std::is_same<T, bool>::value || std::is_same<T, char>::value
-               || std::is_same<T, unsigned char>::value || std::is_same<T, short>::value
-               || std::is_same<T, unsigned short>::value || std::is_same<T, int>::value
-               || std::is_same<T, unsigned int>::value || std::is_same<T, long long>::value
-               || std::is_same<T, unsigned long long>::value
-               || std::is_same<T, rocprim::int128_t>::value
-               || std::is_same<T, rocprim::uint128_t>::value
-               || std::is_same<T, rocprim::half>::value || std::is_same<T, float>::value
-               || std::is_same<T, rocprim::bfloat16>::value || std::is_same<T, double>::value;
+        return ::rocprim::detail::variadic_list<
+            bool,
+            char,
+            unsigned char,
+            short,
+            unsigned short,
+            int,
+            unsigned int,
+            long long,
+            unsigned long long,
+            float,
+            double,
+            rocprim::int128_t,
+            rocprim::uint128_t,
+            rocprim::half,
+            rocprim::bfloat16,
+            __hip_bfloat16>::any([](auto t) { return std::is_same_v<decltype(t), T>; });
     }
 
     /// \brief If `T` is fundamental type, then returns `false`.
@@ -1270,6 +1278,24 @@ struct traits::define<double>
 template<>
 struct traits::define<rocprim::bfloat16>
 {
+    /// \brief Trait `is_arithmetic` for this type
+    using is_arithmetic = traits::is_arithmetic::values<true>;
+    /// \brief Trait `number_format` for this type
+    using number_format
+        = traits::number_format::values<traits::number_format::kind::floating_point_type>;
+    /// \brief Trait `float_bit_mask` for this type
+    using float_bit_mask = traits::float_bit_mask::values<uint16_t, 0x8000, 0x7F80, 0x007F>;
+};
+
+/// \brief This is the definition of traits of `__hip_bfloat16`
+/// HIP arithmetic type
+template<>
+struct traits::define<__hip_bfloat16>
+{
+    // Note: __hip_bfloat16 is included from <hip/hip_bf16.h> whilst rocprim::bfloat16
+    // (hip_bfloat16) is included from <hip/hip_bfloat16.h>. These two types are
+    // distinctly different!
+
     /// \brief Trait `is_arithmetic` for this type
     using is_arithmetic = traits::is_arithmetic::values<true>;
     /// \brief Trait `number_format` for this type

--- a/projects/rocprim/rocprim/include/rocprim/type_traits_functions.hpp
+++ b/projects/rocprim/rocprim/include/rocprim/type_traits_functions.hpp
@@ -1,4 +1,4 @@
-// Copyright (c) 2017-2025 Advanced Micro Devices, Inc. All rights reserved.
+// Copyright (c) 2017-2026 Advanced Micro Devices, Inc. All rights reserved.
 //
 // Permission is hereby granted, free of charge, to any person obtaining a copy
 // of this software and associated documentation files (the "Software"), to deal
@@ -273,8 +273,7 @@ struct is_tuple_of_references<::rocprim::tuple<Args...>>
 private:
     template<size_t Index>
     ROCPRIM_HOST_DEVICE
-    static constexpr
-        typename std::enable_if<(Index < sizeof...(Args)), bool>::type is_tuple_of_references_impl()
+    static constexpr std::enable_if_t<(Index < sizeof...(Args)), bool> is_tuple_of_references_impl()
     {
         using tuple_t   = ::rocprim::tuple<Args...>;
         using element_t = ::rocprim::tuple_element_t<Index, tuple_t>;
@@ -283,7 +282,7 @@ private:
 
     template<size_t Index>
     ROCPRIM_HOST_DEVICE
-    static constexpr typename std::enable_if<(Index == sizeof...(Args)), bool>::type
+    static constexpr std::enable_if_t<(Index == sizeof...(Args)), bool>
         is_tuple_of_references_impl()
     {
         return true;
@@ -543,14 +542,33 @@ template<typename T>
 struct word_type<const volatile T> : word_type<T>
 {};
 
-} // namespace detail
-
-namespace detail
-{
 template<typename Destination, typename Source>
 constexpr bool is_valid_bit_cast
     = sizeof(Destination) == sizeof(Source) && std::is_trivially_copyable<Destination>::value
       && std::is_trivially_copyable<Source>::value;
+
+/// \brief Utility to apply functions over multiple types.
+template<class... Ts>
+struct variadic_list
+{
+    /// \brief Applies predicate `f` over `Ts` and returns if any is true.
+    template<class F>
+    static constexpr bool any(F f)
+    {
+        return (f(Ts{}) || ...);
+    }
+
+    /// \brief Applies predicate `f` over `Ts` type and returns if all is true.
+    template<class F>
+    static constexpr bool all(F f)
+    {
+        return (f(Ts{}) && ...);
+    }
+};
+static_assert(variadic_list<short, int>::all([](auto t)
+                                             { return std::is_integral_v<decltype(t)>; }));
+static_assert(variadic_list<float, int>::any([](auto t)
+                                             { return std::is_integral_v<decltype(t)>; }));
 } // namespace detail
 
 END_ROCPRIM_NAMESPACE

--- a/projects/rocprim/rocprim/include/rocprim/types.hpp
+++ b/projects/rocprim/rocprim/include/rocprim/types.hpp
@@ -1,4 +1,4 @@
-// Copyright (c) 2017-2025 Advanced Micro Devices, Inc. All rights reserved.
+// Copyright (c) 2017-2026 Advanced Micro Devices, Inc. All rights reserved.
 //
 // Permission is hereby granted, free of charge, to any person obtaining a copy
 // of this software and associated documentation files (the "Software"), to deal
@@ -23,6 +23,8 @@
 
 #include <type_traits>
 
+#include <hip/hip_bf16.h>
+#include <hip/hip_bfloat16.h>
 #include <hip/hip_vector_types.h>
 
 // Meta configuration for rocPRIM

--- a/projects/rocprim/scripts/copyright-date/check-copyright.sh
+++ b/projects/rocprim/scripts/copyright-date/check-copyright.sh
@@ -73,6 +73,9 @@ fi
 # Current year
 year="$(date +%Y)"
 
+# Change working directory to git root
+cd "$(git rev-parse --show-toplevel)"
+
 # Enable rename detection with full matches only, this skips copyright checks for file name only
 # changes.
 diff_opts=(-z --name-only '--diff-filter=MA' '--find-renames=100%')

--- a/projects/rocprim/test/rocprim/test_radix_key_codec.cpp
+++ b/projects/rocprim/test/rocprim/test_radix_key_codec.cpp
@@ -1,6 +1,6 @@
 // MIT License
 //
-// Copyright (c) 2024-2025 Advanced Micro Devices, Inc. All rights reserved.
+// Copyright (c) 2024-2026 Advanced Micro Devices, Inc. All rights reserved.
 //
 // Permission is hereby granted, free of charge, to any person obtaining a copy
 // of this software and associated documentation files (the "Software"), to deal
@@ -57,6 +57,12 @@ std::ostream& operator<<(std::ostream& os, const extract_digit_params& params)
     sstream << "{ start: " << params.start << ", radix_bits: " << params.radix_bits
             << ", expected_result: 0x" << std::hex << params.expected_result << " }";
     return os << sstream.str();
+}
+
+inline std::ostream& operator<<(std::ostream& stream, const __hip_bfloat16& value)
+{
+    stream << static_cast<float>(value);
+    return stream;
 }
 
 class RadixKeyCodecTest : public ::testing::TestWithParam<extract_digit_params>
@@ -321,7 +327,8 @@ using TypedRadixKeyCodecTestTypes
                        TypedRadixKeyCodecTestParams<rocprim::half>,
                        TypedRadixKeyCodecTestParams<rocprim::bfloat16>,
                        TypedRadixKeyCodecTestParams<float>,
-                       TypedRadixKeyCodecTestParams<double>>;
+                       TypedRadixKeyCodecTestParams<double>,
+                       TypedRadixKeyCodecTestParams<__hip_bfloat16>>;
 
 TYPED_TEST_SUITE(TypedRadixKeyCodecTest, TypedRadixKeyCodecTestTypes);
 

--- a/projects/rocprim/test/rocprim/test_type_traits_interface.cpp
+++ b/projects/rocprim/test/rocprim/test_type_traits_interface.cpp
@@ -1,4 +1,4 @@
-// Copyright (c) 2024-2025 Advanced Micro Devices, Inc. All rights reserved.
+// Copyright (c) 2024-2026 Advanced Micro Devices, Inc. All rights reserved.
 //
 // Permission is hereby granted, free of charge, to any person obtaining a copy
 // of this software and associated documentation files (the "Software"), to deal
@@ -138,8 +138,12 @@ class RocprimFloatingPointTests : public ::testing::Test
 public:
     using input_type = Params;
 };
-using FloatingPointTypeTestParams = ::testing::
-    Types<rocprim::half, rocprim::bfloat16, float, double, type_traits_test::custom_float_type>;
+using FloatingPointTypeTestParams = ::testing::Types<rocprim::half,
+                                                     rocprim::bfloat16,
+                                                     float,
+                                                     double,
+                                                     type_traits_test::custom_float_type,
+                                                     __hip_bfloat16>;
 
 template<class Params>
 class RocprimIntegralTests : public ::testing::Test

--- a/projects/rocprim/test/rocprim/test_utils_data_generation.hpp
+++ b/projects/rocprim/test/rocprim/test_utils_data_generation.hpp
@@ -365,8 +365,9 @@ auto make_distribution(U min, V max)
 {
     if constexpr(rocprim::is_floating_point<T>::value)
     {
-        using dis_type = typename std::conditional<std::is_same<rocprim::half, T>::value
-                                                       || std::is_same<rocprim::bfloat16, T>::value,
+        using dis_type = typename std::conditional<std::is_same_v<rocprim::half, T>
+                                                       || std::is_same_v<rocprim::bfloat16, T>
+                                                       || std::is_same_v<__hip_bfloat16, T>,
                                                    float,
                                                    T>::type;
         return std::uniform_real_distribution<dis_type>(static_cast<dis_type>(min),

--- a/projects/rocprim/test/rocprim/test_utils_data_generation.hpp
+++ b/projects/rocprim/test/rocprim/test_utils_data_generation.hpp
@@ -1,4 +1,4 @@
-// Copyright (c) 2021-2025 Advanced Micro Devices, Inc. All rights reserved.
+// Copyright (c) 2021-2026 Advanced Micro Devices, Inc. All rights reserved.
 //
 // Permission is hereby granted, free of charge, to any person obtaining a copy
 // of this software and associated documentation files (the "Software"), to deal
@@ -147,6 +147,37 @@ public:
         return bfloat16(std::numeric_limits<float>::signaling_NaN());
     };
 };
+
+template<>
+struct numeric_limits<__hip_bfloat16> : std::numeric_limits<__hip_bfloat16>
+{
+public:
+    static inline __hip_bfloat16 max()
+    {
+        return __hip_bfloat16(std::numeric_limits<float>::max() * 0.998);
+    };
+    static inline __hip_bfloat16 min()
+    {
+        return __hip_bfloat16(std::numeric_limits<float>::min());
+    };
+    static inline __hip_bfloat16 lowest()
+    {
+        return __hip_bfloat16(std::numeric_limits<float>::lowest() * 0.998);
+    };
+    static inline __hip_bfloat16 infinity()
+    {
+        return __hip_bfloat16(std::numeric_limits<float>::infinity());
+    };
+    static inline __hip_bfloat16 quiet_NaN()
+    {
+        return __hip_bfloat16(std::numeric_limits<float>::quiet_NaN());
+    };
+    static inline __hip_bfloat16 signaling_NaN()
+    {
+        return __hip_bfloat16(std::numeric_limits<float>::signaling_NaN());
+    };
+};
+
 // End of extended numeric_limits
 
 END_ROCPRIM_NAMESPACE

